core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java

```diff
@@ -64,8 +64,7 @@
     private int queryManagerExecutorPoolSize = 5;
```

Comment:
Is:
```
Additionally, it the high number of threads causes the lock contention
on io.trino.execution.scheduler.PipelinedStageExecution's monitor.
```
the root cause? Do you know which method is it specifically?

It's worrying that we have to cap `remoteTaskMaxCallbackThreads` value. It means there is more work than threads. It also means that tasks will pile-up leading to system slowdown.

Did you try tracing what takes the most time in these threads. Maybe https://github.com/trinodb/trino/issues/10842 would help

Reply:
do `remoteTaskMaxCallbackThreads` threads do any significant IO? I see `io/trino/server/remotetask/HttpRemoteTask.java:638` but this is async so it won't block the thread.

If there is no IO, having the number of threads significantly higher than CPU count can only drop performance due to context switching and locks so this looks like a good change.

Reply:
`io.trino.server.remotetask.HttpRemoteTask#scheduleUpdate` might trigger multiple executor tasks, each of them will block on `sendUpdate` (apart one). Hence, this way, thread pool can be exhausted.

Reply:
> do `remoteTaskMaxCallbackThreads` threads do any significant IO? I see `io/trino/server/remotetask/HttpRemoteTask.java:638` but this is async so it won't block the thread.
> 
> If there is no IO, having the number of threads significantly higher than CPU count can only drop performance due to context switching and locks so this looks like a good change.


Yes, I've analyzed the usage of that `executor`  and I did not see the blocking I/O. I saw the one you mentioend but, as you said, it is async. Anyway, even if we spot any blocking I/O then we should rethink this approach.

Reply:
> io.trino.server.remotetask.HttpRemoteTask#scheduleUpdate might trigger multiple executor tasks, each of them will block on sendUpdate (apart one). Hence, this way, thread pool can be exhausted.

Yes, exactly. But why it is a problem? We have limited thread pool so it could be always fully saturated. But, why we need 1000 threads, not 500, not 2000?

Reply:
>  It means there is more work than threads. It also means that tasks will pile-up leading to system slowdown.

On my eye, the number of threads in CPU-bound workload is not important, especially to high number of them - I mean.

Reply:
I guess, one task could block other tasks.

Reply:
FYI: Before merging that I need to wait for fixed version of our internal tool to measure throughput (benchto).
For now, it is not possible to use that because of the bug. I am using the old way to compute throuhgput but it looks like it works correctly as well, especially as a tool to compare two different version, not to get referal value for cluster throughput.